### DISCUSSION TITLE
Prioritize Modernization investments

### DISCUSSION THREAD

**Discussion by pchaurasia14** [dotnet/wpf#7555] [Announcements]
We kindly ask you for your valuable input on the priorities for modernizing investments that the WPF team should focus on, as outlined in the [roadmap document for 2023](https://github.com/dotnet/wpf/blob/main/roadmap.md#modernizing-wpf).

Please use thumbs up (no other emoji responses will be considered to avoid duplication) on options below which you would like us to prioritize.

**Comment by pchaurasia14** [comment]
1. Windows 11 Theming

| Goal | Description | Rationale |
| ------------- |:--------------|:-------------|
| Windows 11 Theming | Bringing Windows 11 look and feel for majority of WPF controls. <br/> Support for Win11 features such as snap layout, rounded corners for controls and newer color schemes would bring enhanced experience for WPF applications. <br/> We will iterate on full scope of the work with WPF community. | For all consumer applications that are built on WPF running on Win11, this feature would ensure that applications can take advantage of modern design elements and behaviors. |

**Reply by SkinnySackboy** [reply]
This would be great, including message boxes and file/folder open/save dialogs, as currently they look pretty dated even if you use something like MahApps to make the rest of the application look modern.

**Reply by 292390450** [reply]
I am happy to have an update, but I think it is more important for DX updates.

**Reply by pchaurasia14** [reply]
@292390450 - DX updates are not currently planned for this fiscal year. We will revisit the plan once we have better clarity.

**Reply by jschroedl** [reply]
If this becomes a 2023 priority, do you anticipate that it will include support for the more desirable compact layout instead of the standard (large) margins/padding/spacing?  Our app is information dense and needs much tighter, more compact, spacing as is currently available with WinUI with the compact layout.  Ref: [Fluent Compact sizing](https://learn.microsoft.com/en-us/windows/apps/design/style/spacing#fluent-compact-sizing)

**Reply by pchaurasia14** [reply]
@jschroedl - We haven't planned on integrating the compact layout. We can certainly consider it if there is interest in the community.

**Reply by ChrisBaker812** [reply]
I can't believe it's come to this. Voting for three fundamental pieces of functionality WPF should have had 5+ years ago with ongoing investment, and, to add insult to already serious injury, we may only get one of them! And all we hear is excuses about small team and team maturity. MS, have you forgotten who it was that helped make the company what it is today? That's developers by the way, because Windows does nothing without it's software and the majority of that over the years was Desktop software, written using Winforms and then WPF. Now we're considered orphans of the development world, having to beg for the gruel of new functionality like Oliver, tentatively asking for more for a product that I love. We still exist (desktop developers) not everything has to be (or can be) web or mobile. Both WinUI 3 and MAUI, are not only unfinished, they are a cruel joke to make Winforms/WPF (or rather desktop) developers believe there is a future for them in the MS development product world.

**Reply by SteveDunn** [reply]
Eloquently put!

**Reply by pocodicom** [reply]
Not sure I understand this feature. WPF controls can already be styled anyway you want. Is this just another style theme such as the Aero theme?

**Reply by pchaurasia14** [reply]
@pocodicom - Yes. it will be a similar style theme

**Reply by jschroedl** [reply]
This feature would implement the look and feel of the Windows 11 controls as a built-in WPF theme. We should not have every WPF developer having to recreate the OS visuals themselves or via a library such as Telerik, MahApps, etc. People will get it slightly wrong, will cut corners, etc. and the results will be piecemeal. Not having this built into WPF today is one of the reasons why users of WPF apps can rightfully complain about the app not feeling modern or looking like a proper desktop app compared against the apps packaged with the OS.   You are obviously correct that controls can already be styled and this is good but also difficult to get right. I would prefer to have my tweaks based on the MS-provided theme than my own cobbled together attempt. I have better things to spend time on than recreating the OS theme.

**Reply by wstaelens** [reply]
I agree @jschroedl rather have all visuals the same OS style instead of Telerik, DevExpress, ... each having their own one slightly different compared to the Windows 10/11 OS.

How many people are in the Microsoft WPF team who work on this?
Can't you do this:

- 50% of MS WPF Team: work on old/long term issues/frustrations
- 50% of MS WPF Team: work on new features (controls, themes, whatever...)
- Community: focus on performance updates, code review, code improvements, less allocations ... 

MS team has best knowledge on the internal workings of the WPF libraries.
The community can easily improve small portions of code, replace arraylists, improve allocations, and small PR can be easily reviewed/accepted/tested by MS WPF team (e.g. every friday).

Everybody happy.

**Reply by pocodicom** [reply]
Two more questions:
Would this theme support both light and dark mode?
Will this theme be backward compatible with .NET Framework 4.8?

**Reply by pchaurasia14** [reply]
@wstaelens - Thank you for your suggestion! We agree that it's a great idea to have a balance between working on old/long term issues and developing new features. However, the work that is not tracked on GitHub (such as servicing and security fixes) do take up team's bandwidth. 
We also agree that the community can play a vital role in improving the performance and code quality of WPF. We will definitely consider incorporating this approach in our future dealings. 

As for the number of people in the WPF team, we cannot disclose that information, but rest assured that we have a dedicated team working on WPF.

**Reply by pchaurasia14** [reply]
@pocodicom - 
1. We are considering support for both light and dark themes. 
2. We will not be backporting it to .NET Framework.

**Reply by uecasm** [reply]
That's disappointing.  I have a very large number of WPF apps that cannot migrate from Framework, and it will look a bit weird if they are the only things not Win11 themed.

Ideally, you should work with the native designers, and the OS theme should simultaneously come out for Windows itself as well as both versions of WPF.  That should just be the core deliverables of the theme.  If you can't, then you should reduce the frequency of changing the theme until you can.

**Reply by aquinn39** [reply]
Perhaps as part of modernising the UI themes, https://github.com/dotnet/wpf/issues/5936 and https://github.com/dotnet/wpf/issues/5937 could be looked into as they are in relation to providing modern scrolling behaviour and so would fit nicely with a modern Windows 11 style theme. And maybe https://github.com/dotnet/wpf/issues/5938 and https://github.com/dotnet/wpf/issues/5939 as well? These would all certainly improve the user experience plus make WPF apps behave in a more modern way and fit in better with Windows 10 and 11.

**Reply by danmoseley** [reply]
> That's disappointing. I have a very large number of WPF apps that cannot migrate from Framework, and it will look a bit weird if they are the only things not Win11 themed.

just curious (I don't work on WPF) what are the reasons you cannot migrate the apps? Of course there are potentially good reasons, eg., it is not free. But I'm wondering whether it's breaking changes, incompatible dependencies, possibly lifecycle, etc.

**Reply by maxima120** [reply]
> That's disappointing

There is a very well known wpf controls suite that costs $900. It has 60 themes that you can apply on out of the box microsoft controls easily. If you can't afford that may be you shouldnt be voting for features, skewing the actual developers representation.

**Reply by kevwkev** [reply]
Any chance we could get a modern Office 365 Ribbon control and Simplified Ribbon style?

<img width="775" alt="Screenshot 2023-05-17 at 12 57 26 PM" src="https://github.com/dotnet/wpf/assets/1664903/bc4c0545-540c-4a99-a29c-8c8887ad1f71">

**Reply by NeverMorewd** [reply]
Well, It seems that Windows 11 Theming win the election.So have we started yet?

**Reply by pchaurasia14** [reply]
@NeverMorewd - Soon enough. There are some things pending on the internal partner dependencies. Once sorted, we should be able to make progress on this front.

**Reply by PabloGallegosH** [reply]
Hello good day Every one. Excelent job.  Starter now this plan?

**Comment by pchaurasia14** [comment]
2. Newer Controls

| Goal | Description | Rationale |
| ------------- |:--------------|:-------------|
| Newer controls | [WPF FolderBrowserDialog](https://github.com/dotnet/wpf/issues/438) - Introducing native support for FolderBrowserDialog for WPF | This has been a top ask from the community since .NET Core 3. This feature would reduce dependency on WinForms and other third-party alternatives. |

**Reply by JaykeBird** [reply]
When discussing "new controls", are you only considering FolderBrowserDialog here? Or are there other new controls on the table too, such as NumericUpDown, TimePicker, and/or maybe porting some controls over from WinUI (just to give examples)?

**Reply by pchaurasia14** [reply]
@JaykeBird - Yes, for now, we plan on supporting folderbrowserdialog. But will be happy to introduce more controls based on community interests.

**Reply by deitry** [reply]
+1 for TimePicker

**Reply by uecasm** [reply]
I've always found a folder browser to be completely useless, because it doesn't allow you to type in a path.  I always use a plain `OpenFileDialog` with a dummy filename+filter instead; works much better.  There's other new controls I would like, but that one definitely isn't one of them.

**Reply by aquinn39** [reply]
> I've always found a folder browser to be completely useless, because it doesn't allow you to type in a path.  I always use a plain `OpenFileDialog` with a dummy filename+filter instead; works much better.  There's other new controls I would like, but that one definitely isn't one of them.

That's the old FolderBrowser. The newer one is much better - it looks and behaves more like a file browser including letting users type in a path.

**Reply by Kurpanik** [reply]
With [ookii. dialogs](https://github.com/ookii-dialogs/ookii-dialogs-wpf) there ist already a WPF solution, which solves the problem of the missing dialogue very quickly. For the other two options, however, there is far less possibility of implementation.

**Reply by AlexGoOn** [reply]
Thank you for considering implementing this feature. If you decide to move forward with it, I would like to suggest that you also consider the possibility of adding support for custom theming to the dialog window. Doing so would provide users with a more cohesive and personalized experience, as they would be able to customize the look and feel of the dialog to match the rest of their application.

Probably, supporting custom theming will be available without additional investments on your side, however, I wanted to make sure that you are keeping this in mind as you move forward with the implementation.

**Reply by harikrishnan-p-v** [reply]
Thanks for implementing this I have been using "Microsoft-WindowsAPICodePack-shell" by rpastric,contre,dahall. Introducing native support for FolderBrowserDialog for WPF is a much awaited feature.

**Comment by pchaurasia14** [comment]
3. Nullability annotations

| Goal | Description | Rationale |
| ------------- |:--------------|:-------------|
| Nullability annotations | Enable nullability annotations in WPF | This increases the quality of the code base, as well as the quality for all WPF apps consuming it and reduces time spent debugging `ArgumentNullException`s and `NullReferenceException`s. <br/>Rest of the dotnet (eg. winforms) repo is already moving in the direction and this goal would bring WPF to latest standards as well. <br/> We welcome any community contributions in this area.  |

**Reply by danmoseley** [reply]
@pchaurasia14 for this one, are you aware of any work that only the WPF team is equipped to do, or is this open for contributions today? in dotnet/runtime, some community folks were a huge help with nullable, @maxkoshevoi and @stevedunn I remember did a lot, also I think @huoyaoyuan  @hughbe and others. Perhaps some of those folks may be interested in helping annotate WPF.

edit: what I think helped was the team created issues with checklists, one per assembly, walking up the assembly dependencies in groups, so folks could pick up work without overlap. [like this](https://github.com/dotnet/runtime/issues/43605) and [this](https://github.com/dotnet/runtime/issues/41720). perhaps the WPF team might do that.

**Reply by dotMorten** [reply]
My suggestion would be to create a list of assemblies in priority, starting with the lowest dependency in the list (like `System.Xaml` for instance), and working your way up. Typically how I've done it is explicitly disable null annotations in every single file in a project, then enable it project wide. Then you chip away at it by removing the disable in a file, work through the null annotation warnings (which should be set to be compile errors). It's easy to make a small tool that counts number of lines of code per file and whether the annotation is enabled or disabled, and that gives you a great progress report, and idea of how far each project is.
I think this could very much be done as a community project, but I think we need the team to do the early planning work and organizing priorities and track progress. It's a bit of an undertaking and might span a few releases, so you could set a goal of X set of projects per release.

**Reply by JaykeBird** [reply]
In my personal opinion, this one should definitely be a focus - nullability annotations is a great QoL feature that the rest of the .NET repos are implementing but WPF still needs to catch up on.

That being said, I know that more visual and apparent changes, such as new controls or updated theming, are definitely important too, to help buck the (mis-)perceived sentiment that "WPF is dead". But in my opinion, nullability annotations is something that should be added in sooner rather than later.

**Reply by pchaurasia14** [reply]
@danmoseley @dotMorten - Great idea, thanks for suggestions. I just created new [issue](https://github.com/dotnet/wpf/issues/7563) and we will populate the plan there in a few weeks. It is great if there is interest from community to contribute, we would be thrilled to collaborate on that and we will make sure to make time for PR reviews and project driving.

**Reply by zspitz** [reply]
This seems to me the lowest hanging fruit:

1. The need can't be filled by third-party libraries (unlike [file dialogs](https://github.com/ookii-dialogs/ookii-dialogs-wpf)).
2. It requires the least amount of domain internals knowledge (unlike updates to theming).
3. Each individual change has a specific, relatively local effect, thus making it easier for the community to collaborate.

**Reply by SteveDunn** [reply]
I'd be happy to work on some of these annotations. @zspitz sums it up perfectly; I don't have much knowledge of the internals, but would be happy to skip around the code to determine nullability of existing contracts.

**Reply by kant2002** [reply]
https://github.com/dotnet/winforms/pulls?q=is%3Apr+is%3Aopen+nullability Here basically one man show for supporting nullable annotations. It was done in small steps. Basically you need to show support for some contributor and process will happens by itself. Probably it can be done in WinForms because they have tests already in place.

**Reply by maxima120** [reply]
> 3. Nullability annotations
This whole nullability phenomenon is fake. Like the flat earth theory. Made up by incompetent professors from little popular universities in Marxism leaning countries and promulgated by part of devepers community for personal gain or out of ignorance.

**Comment by BlameTwo** [comment]
Is it compatible with higher DX version?

**Reply by ghost** [reply]
+1

**Reply by 292390450** [reply]
1

**Reply by cliyr** [reply]
+1

**Reply by LightBender** [reply]
+1

**Reply by ghost** [reply]
+1

**Reply by NeverMorewd** [reply]
+1

**Reply by lindexi** [reply]
Can we get something from [D3D9On12](https://github.com/microsoft/D3D9On12) ?

**Reply by SkyIsWhite** [reply]
+1

**Comment by wstaelens** [comment]
https://github.com/dotnet/wpf/issues/3546 
https://github.com/dotnet/runtime/pull/78374#issuecomment-1432663604 
± 7 - 8 years and still not fixed. 🤞

**Reply by batzen** [reply]
@wstaelens could you please stop mentioning the XPS issues in every conversation that comes up?
This discussion is a vote for something totally unrelated to XPS.

**Reply by wstaelens** [reply]
@batzen fixed in .net 9 preview 7 ;-)

**Reply by h3xds1nz** [reply]
And it was a runtime issue as you sure have noticed.

**Reply by wstaelens** [reply]
@h3xds1nz xps is a subset of xaml and indeed the culprit was located in the compression libraries. But improvements are coming towards  XPS which potentially might improve XAML/WPF also: https://github.com/dotnet/wpf/issues/4000#issuecomment-2266642316

**Comment by SkinnySackboy** [comment]
If you're looking at DX improvements perhaps the WinForms airspace issue could be resolved, as it's something that still causes us headaches?

**Reply by pchaurasia14** [reply]
@SkinnySackboy  - DX improvements are currently not planned for this year. We will revisit the plan for future releases.

**Reply by SkinnySackboy** [reply]
Ok thanks.

**Reply by johnottone** [reply]
> If you're looking at DX improvements perhaps the WinForms airspace issue could be resolved, as it's something that still causes us headaches?

@SkinnySackboy, It's funny I ran across this post not long ago. Not saying it is going to solve any problems you maybe having or anything, but you never know, it may spark an idea or something, that may lead to you discovering a solution...

[jonathanantoine - WPF 4.5 – Part 8: No more airspace problems - Integrating WPF with win32: finally, I can use the WebBrowser control](http://www.jonathanantoine.com/2011/09/24/wpf-4-5-%e2%80%93-part-8-no-more-airspace-problems-integrating-wpf-with-win32/)

[LOTS of USEFUL KNOWLEDGE](https://dwayneneed.github.io/wpf/2013/02/26/mitigating-airspace-issues-in-wpf-applications.html)

Check out this control and get some ideas from it perhaps:
[Airspace Fixer](https://github.com/chris84948/AirspaceFixer)

and also there is this which youve probably read, but still relevant:

[WPF Interoperation](https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-3.0/aa970688(v=vs.85))

[Changing Layout Dynamically in Windows Forms](https://www.developer.com/microsoft/dotnet/changing-layout-dynamically-in-windows-forms/)


also, see TableLayoutPanel, FlowLayoutPanel, Docking controls,

**Comment by StefanKoell** [comment]
While I understand the desire for new controls, I think the priorities and the focus should be at a fundamental level. Areas which are not (and will not be) covered by 3rd parties, like 

- more efficient and faster rendering (new DirectX version, better call batching, etc)
- improved interop (airspace issues)
- compiled bindings
- native SVG support

To name just a few. There have been lots of issues/discussions over the year but no progress at all.

**Reply by 292390450** [reply]
Indeed, its rendering performance now seems to be a bit slow compared with avalonia，More updates such as reflective performance and virtualized ItemsRepeater make wpf look faster

**Reply by pchaurasia14** [reply]
@StefanKoell - I agree that areas that are not covered by 3rd party solutions should be our focus. However, some of those areas are lacking adequate test coverage which puts us at a disadvantage since we cannot ascertain if the newer changes will cause unexpected regressions. 
Which is why our plan is to first open source the tests and then work on bigger initiatives.

**Comment by pvillads** [comment]
1. Theming
2. Nullability support

**Reply by pchaurasia14** [reply]
@pvillads - Thanks for the feedback. Appreciate if you could cast your vote via thumbs up on the above items.

**Comment by jimfoye** [comment]
I'm going to check in and see how you guys are doing in 2024. Oh wait, I'm retiring that year.

**Comment by rbqm** [comment]
Whatever you do, please do not break backward compatibility with .NET Framework. For mission critical WPF applications with release cycles measure in years it is impossible to migrate from .NET Framework to .NET X because of the difference in lifecycle policy between .NET Framework and .NET X. The new .NET X lifecycle policy where version X receives security updates for a limited amount of time only is not suitable for such applications, unless there is a way to upgrade applications installed in the field from .NET X to .NET X+1 without having to release a new version of the application itself. I hope that Microsoft realizes that there needs to be a future version of .NET X that has the same lifecycle policy as .NET Framework. When/if that time comes, I want to be able to port my application from .NET Framework to .NET X without having to rewrite it.

**Reply by pchaurasia14** [reply]
@rbqm - Absolutely. We are committed to ensure our backward compatibility promise remains unbroken. As for automatic upgrades to newer .NET versions without application release, it needs to be planned at .NET level. AFAIK, there is nothing planned for it yet. But you may consider opening an issue in https://github.com/dotnet/runtime repo.

**Reply by jimfoye** [reply]
@rbqm This is just my opinion, but I think MS is increasing the major version number as fast as possible to encourage classic framework developers to upgrade. I.e. you will feel more "behind" if you are targeting 4 and .NET Core is already up to 8 (which it will be in November). Ironically, this actually deters upgrades, for the reasons you point out.

**Reply by johnottone** [reply]
I don't really think that MS ups the version for that reason... They have lifecycle planning and standardized release dates on most products...I thought... And also, I can't say that **"feeling behind" is what deters upgrades**. Sometimes, a simple update is not quite as simple as it seems. Upgrading an app from .NET 4.72 up to .NET8, requires a user to have permissions to install the runtime. That is the deterrence-- That upgrade also requires tedious cooperation with other companies that handle user permissions and/or group policy. And the most complex upgrades deal with ASP.NET because the packages change in the Visual Studio templates, and on a 5 year old website that used packages.config, where I'm just trying to add a couple fields to a form, I am pretty much forced to do an upgrade because of breaking changes in the underlying templates. That being said, with the promise of backward compatibility, and the nature of how WPF controls can be themed/styled I don't think that changing the version is even required. Thats kind of why I pitched the ThemeBuilder, which could require actually selecting the framework version to modernize, which loads the known control templates for that .Net version and would attempt to normalize it and add to it, and of course validate against their version, so that in the end everyone would end up with a theme or style that is "modernized" yet compatible across the board. Maybe I'm off bas a little, but that was my take on it.

--think lowest common denominator

**Reply by Teddy55Codes** [reply]
> Upgrading an app from .NET 4.72 up to .NET8, requires a user to have permissions to install the runtime.

https://learn.microsoft.com/en-us/dotnet/core/deploying/
You can simply build your project as self-contained. I did that with my last WPF app, which was written in .NET 8.

**Reply by jimfoye** [reply]
@Teddy55Codes How big was it?

**Reply by Teddy55Codes** [reply]
I think it was some where around 60MB bigger

**Reply by jimfoye** [reply]
So let's say your app is 10MB - that's 70MB, you're installing the entire .NET runtime on the computer in order to run your application. Meanwhile, other folks are doing the same, you have to wonder how many .NET runtimes a computer is going to end up having installed on it? It's good to have the option, of course, but many of us want to do things "the old way" where the runtime is installed on the computer and any app compiled for that runtime will work. Microsoft does not seem to see things that way, though.

**Reply by Teddy55Codes** [reply]
I do see your perspective. However, .NET is no longer a Windows only technology. In my opinion that is actually very good. Things like the GAC where always a head scratcher for me. I happily take a bit of a larger app size if it means i have to deal with less Windows bs.

**Reply by jimfoye** [reply]
> I do see your perspective. However, .NET is no longer a Windows only technology. 

Yeah but we're talking about WPF apps here. In that context, it is a Windows only technology. (And always will be, apparently).

**Comment by pchaurasia14** [comment]
Thank you very much for your insights folks and for your help modernization investments prioritization!

As of 2/3/2023, we have current voting status:

- Windows 11 Theming - 225 votes
- Newer Controls - 155 votes
- Nullability Support - 99 votes 

Based on the results, we are inclined to channel our focus on delivering on the Windows 11 theming and newer controls support as top 2 items.
As for Nullability support -- we are thankful for community interest in helping with the work. We are actively finalizing the plan (draft plan is already available [here](https://github.com/dotnet/wpf/issues/7563)) to guide community contributions in the space. We plan to make PR reviews for Nullability support our top priority to unblock the community effort. 

Thank you! And as usual, let us know what you think or if we missed anything.

**Reply by Symbai** [reply]
> let us know what you think or if we missed anything.

I honestly looked for AOT and trimming support and to my surprise it wasn't even available for voting (and the PRs for that are still not reviewed ofc). A self contained WPF app is over 200mb and it won't get any better. Windows 11 theming? There are toooons of good UI libaries out there giving you that and more, same for new controls. I really think its not worth the effort and time to work on something which can be done already. Instead give us something we cannot do ourselves and we have to rely on your team adding support for it.

**Reply by jimfoye** [reply]
Agreed. And, next year Microsoft will again change the look of Windows somehow and then W11 theming will seem pointless. We'll always be behind. Also, some of us don't even like W11 theming!

**Reply by johnottone** [reply]
I understand that the Windows 11 theming had the highest votes. Why dont we consider something more along the lines of a theme builder that allows the community to select ANY GIVEN FRAMEWORK .Net 3.5, .Net4.7.2, etc... and generate a theme based on the selected framework version and the default template (generic.xaml) that was available for those controls, in that specific time period.  Meanwhile, creating a new theme for Windows 11....AND Windows 10. I mean WPF is supposed to bring us a consistent programming model, and that model SHOULD be consistent with the OS in which it must run. Alot of small companies are having to use 3rd party controls because of all the gaps in WPF's ["comprehensive set"](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/overview/?view=netdesktop-7.0#:~:text=Welcome%20to%20the,the%20.NET%20API.) of tools. Are you aware that Telerik(Progress) offers a PRO Version of their Theme Builder and charges $100.00 for it even to an ULTIMATE SUBSCRIBER with the most expensive package. Small companies don't have the time nor the resources needed to fill in the gaps... So they have to go out there and find other 3rd party dependencies. They might only use a few of the controls like the DataGrid, Navigation(Hamburger)-style Menus, ReportViewers, etc... Just like the File Dialog, the set is missing a lot of key controls and features. Ive been using WPF since 2007. 15 Years have passed since I started using WPF. **And we are just now talking about THEMES that are consistent with the OS???**

**Reply by Symbai** [reply]
Well, at least they CAN use 3rd party controls and not even less of them. I'm waiting for AOT support since it was announced (with .NET 3.1 i think). I dont have alternatives. Same for trimming support. And it is extremely obvious that the WPF repo team has incredibly limited time to work on things here. So doing something which has been done before is not only absolut waste of time but also extremely contra productive. It means almost everything else has to stand back, for a long time.

Btw: the real reason why companies switch to (paid!) 3rd party providers is not the theming but the controls. Providing a complex end user filter (not only for string) for a datagrid is something you get out of the box. In WPF default you have to write it yourself, takes a very long time. Time is money. This ain't nothing you will get if you vote here on Windows 11 theme. What you get is maybe! Mica (pro tip: You can already get this yourself) but likely only on window background, not on controls. You likely also won't get the reveal brush effect when moving mouse over controls. But you may get the thing nobody asked for when moving cursor over the maximize button that it throws the grid stuff of Windows 11 alias snap layout. Round corners for controls? Yay, you almost got me with this. I can understand this is absolut top priority... oh wait it isnt. Because now THIS is something that you can already do yourself with just adding one line to each control.

However there were only 3 things to vote for. Nullability isnt that much exciting for most people. On new controls only the folder picker was mentioned which is great to have but it feels a bit less especially as folder picker is something that really only matters for some software. With THESE other two choices I can understand that a new theme has won the voting. But I'm not complaining about the winner, I am complaining about really important features were missing to vote on. Features where we have to RELY on WPF team. With AOT and trimming we could have just a 60mb app vs 230mb that is much more performant. And companies don't have to spend thousands of dollars on Obfuscators anymore. On every new .NET major release the .NET runtime team is celebrating AOT as the new hot shit. So where is it? Why it wasn't available for voting?

**Reply by wstaelens** [reply]
We honestly don't care about a Win 10 / 11 theme right now because there are other bugs that have a bigger importance for us. We want to see the list of bugs and issues being fixed which are open for ±8? (10?) years now in .NET Framework / .net .

Why do we go to a third party controls supplier like DevExpress ( https://www.devexpress.com/products/net/controls/wpf/ ) ? 

- Because the default controls are too limited provided by Microsoft. 
- The support (often an answer after 1 hour. In case there is a bug it is often fixed within a week with a hotfix or comes in the next release).

We expect the core/engine of WPF being fast, smooth, bug free (😄 ), ... and the team having a focus on that.

The Office and Windows 11 themes, we can get them from DevExpress: https://www.devexpress.com/products/net/controls/wpf/themes/

**Reply by batzen** [reply]
The WPF core engine is quite fast, it could be faster of course.

But using DevExpress makes your app extremely slow to load and adds a permanent runtime overhead, once you loaded any control from them, even if you are never using any DevExpress control again during process lifetime.

**Reply by wstaelens** [reply]
@batzen well this could be passed on as feedback to DX. But I mean, I agree with some others here that Win 11 themes can be done via external libraries. I think there are more important things to focus on.  
Reducing memory footprint; less allocations, startup time, bugs, etc...

for example: If I remember correctly months ago there were still many ArrayLists in the sources causing a lot of unneeded boxing/unboxing.
But if the community fixes these but the PR's are not being looked at, it also doesn't make sense. 

But I respect the poll/votes. Hope the team will make out the best of it.

**Reply by pchaurasia14** [reply]
I completely agree with the idea of investing in AOT and trimming support. However, we need to be cautious to ensure that we don't unintentionally disrupt any current features while making these changes.

At the moment, we are facing some test gaps that may make it challenging to achieve these goals in a short amount of time without risking significant regression. That's why we are currently focusing on enhancing our test automation and migration capabilities to minimize any potential risks.

In the meantime, we are working to deliver on long-standing requests that are not heavily dependent on existing tests being open source. As outlined in our roadmap document, our long-term plan does include support for AOT, trimming, DX upgrades, etc.

@johnottone - ThemeBuilder sounds like nice idea. How about logging an issue and see how we can bring it to fruition.

@wstaelens - We are working to improve the situation on PR backlog as well.

**Reply by Symbai** [reply]
> we are facing some test gaps

Then work on it first. THen work on AOT. I have no idea what you mean with "In the meantime we are working on". It sounds like your team has incredibly much free time left to do these things. IF that would be the case, okay I would probably don't mind that much about it. But obviously you don't. We are going close to .NET 8 release and since this repo has been made open source all we hear is that you have to build up tests etc. That is taking more than 5(!!) years and still ongoing.

Go and focus on important things. If it takes more than 5 years to set up tests, then your team don't have the resources to do unimportant work.

**Reply by maxima120** [reply]
> Why do we go to a third party controls supplier like devexpress

How much did you pay for. Net? Nothing? And you expect free 1hr support response and a fix tomorrow? What do you smoke? I want it too

**Reply by wstaelens** [reply]
> > Why do we go to a third party controls supplier like devexpress
> 
> How much did you pay for. Net? Nothing? And you expect free 1hr support response and a fix tomorrow? What do you smoke? I want it too

Don't worry about that, we don't smoke and pay our Microsoft bills each month :-)
I didn't say we expect such a fast support but if we pay for it and it happens so fast, we can only enjoy working with such a third party controls supplier in that way.

I look forward to seeing the promises turned into action in the next weeks.

**Reply by johnottone** [reply]
@pchaurasia14 

> @johnottone - ThemeBuilder sounds like nice idea. How about logging an issue and see how we can bring it to fruition.

I will log it under issues I guess, but wouldnt it be better in the Discussions forum under  the "Ideas" category, or is that where Ideas actually go to die? LOL

**Reply by AlexGoOn** [reply]
@batzen 
As a member of the DevExpress WPF team, I would like to let you know that we appreciate your input. Large third-party component suites, such as DevExpress assemblies, can sometimes cause startup performance issues due to the lack of native images in the system. However, we have found that using Ngen, MultiCore JIT, or ReadyToRun can often help address these issues. In fact, using MultiCore JIT and ReadyToRun together can yield even better results in .NET/.NET Core. You can find more information on these techniques in our blog post here: [9 Tips to Reduce WPF App Startup Time](https://community.devexpress.com/blogs/wpf/archive/2023/01/26/9-tips-to-reduce-wpf-app-startup-time.aspx).

We are always striving to improve our products, and reducing startup time is one of our top priorities. To that end, we are currently working on a new optimized theme for our upcoming release that we believe will further enhance performance. 

Thank you again for your feedback. If you decide to give our components another try, please let me know, and I will personally review your requirements and provide you with general guidelines to help you get started as smoothly as possible.

**Reply by batzen** [reply]
@Alexgoon feel free to contact me on Twitter. Your DMs seem to be closed there.

**Comment by huoyaoyuan** [comment]
Fundamental overhauls like porting C++/CLI to C# should be supported better. They can block a series of further follow-ups. Perhaps BCL experts can help in this area.

**Reply by jimfoye** [reply]
Nothing significant like that is going to happen.

**Comment by Coding201212** [comment]
The air space issue is very serious, there’s no good solution to resolve it perfectly for many years.

**Comment by wgebczyk** [comment]
Maybe admit failure of Maui, winui, etc and provide non Win rederers like Avalonia. You did that with Mono "fork" then you can do WPF 2.0 cross platform.

I know, this wish require some strong visionaries dragging it regardless all obstacles that maintainers. Maybe hire someone that would drag it till happy end?

**Reply by jimfoye** [reply]
I have switched over to the "New Outlook" and...it is a progressive web app! Does not seem like a vote of confidence in native Windows desktop development.

**Reply by wstaelens** [reply]
> I have switched over to the "New Outlook" and...it is a progressive web app! Does not seem like a vote of confidence in native Windows desktop development.

don't switch to the "new outlook", doesn't support COM or anything. Breaks a lot of applications.
Switch to it if you want many support calls 😞

**Reply by johnottone** [reply]
As far as Theming WPF is concerned, I know it was voted on and was put on the roadmap, but if there are more important things that need to be addressed, or resources have been or need to be shifted, well, it happens a lot in my world, so I'm used to it. 


> Maybe admit failure of Maui, winui, etc and provide non Win rederers like Avalonia. You did that with Mono "fork" then you can do WPF 2.0 cross platform.
> 
> I know, this wish require some strong visionaries dragging it regardless all obstacles that maintainers. Maybe hire someone that would drag it till happy end?

I have to strongly disagree with "admitting failure". If that were the answer, then perhaps there wouldn't be WPF today at all. Yes, it took time, but if it was just given up on, then you wouldn't be asking for a WPF2.0. 
I personally like to stay on the cutting edge, leveraging new tools and frameworks of the time, because as these tools emerge, other advances are made. And be clear that sometimes it's not .NET that's behind, it may be the underlying technology is just not there yet.  A couple years back I had the need to start using Xamarin, and it was ....ehhhh...... 
So, I was excited when MAUI was announced, especially with Blazor, because it filled many a void and I gladly replaced our Xamarin apps with MAUI apps, and getting ready to implement 3 more entirely new apps with MAUI Admittedly, this held (and still holds) some uncertainties.

For you specifically, it seems that the lack of FULL NativeAOT WASM, due to certain dependencies on the IL, you may be, or may have been deterred. But the fact that it can be embedded in virtually all frameworks; WPF, WinForms, Maui, etc., and the fact that NOW, you can compile that portion into the Web Assembly itself, it's still very promising, IMHO... I think it still increases the storage, and I acknowledge that it may have kept some devs from leveraging the newer technologies (NET Core, MAUI, WinUI). But I'm optimistic and think that it's a path worth following. Showing my age here, but It kind of reminds me of how WPF started and the uncertainties that existed when I started using it, even before its RC1, back when my company was subscribed to the MSDN (on cdroms...lol) and MS would send us the CTPs. I still have that old medical app demo that MS released with the flipping tabs and live heart rate monitors and such. And still have the very first apps I wrote in WPF (Thank you JMAC). It literally changed my whole world.
